src/solvers/dgsem_tree/indicators_2d.jl

```diff
@@ -185,6 +185,100 @@ function (löhner::IndicatorLöhner)(u::AbstractArray{<:Any,4},
+  ContainerShockCapturingIndicator = Trixi.ContainerShockCapturingIndicatorIDP{real(basis)}(0, nnodes(basis), number_bounds)
+
+  cache = (; ContainerShockCapturingIndicator)
+
+  return cache
+end
+
+function (indicator_IDP::IndicatorIDP)(u::AbstractArray{<:Any,4}, semi, dg::DGSEM, t, dt; kwargs...)
+  @unpack alpha = indicator_IDP.cache.ContainerShockCapturingIndicator
+  alpha .= 0.0
+
+  indicator_IDP.IDPPositivity  &&
+    @trixi_timeit timer() "IDPPositivity"  IDP_positivity!(alpha, indicator_IDP, u, dt, semi)
+
+  # Calculate alpha1 and alpha2
+  @unpack alpha1, alpha2 = indicator_IDP.cache.ContainerShockCapturingIndicator
+  @threaded for element in eachelement(dg, semi.cache)
+    for j in eachnode(dg), i in 2:nnodes(dg)
+      alpha1[i, j, element] = max(alpha[i-1, j, element], alpha[i, j, element])
+    end
+    for j in 2:nnodes(dg), i in eachnode(dg)
+      alpha2[i, j, element] = max(alpha[i, j-1, element], alpha[i, j, element])
+    end
+    alpha1[1,            :, element] .= zero(eltype(alpha1))
+    alpha1[nnodes(dg)+1, :, element] .= zero(eltype(alpha1))
+    alpha2[:,            1, element] .= zero(eltype(alpha2))
+    alpha2[:, nnodes(dg)+1, element] .= zero(eltype(alpha2))
+  end
+
+  return nothing
+end
+
+@inline function IDP_positivity!(alpha, indicator_IDP, u, dt, semi)
+
+  # Conservative variables
+  for (index, variable) in enumerate(indicator_IDP.variables_cons)
+    IDP_positivity!(alpha, indicator_IDP, u, dt, semi, variable, index)
+  end
+
+  return nothing
+end
+
+@inline function IDP_positivity!(alpha, indicator_IDP, u, dt, semi, variable, index)
+  mesh, equations, dg, cache = mesh_equations_solver_cache(semi)
+  @unpack antidiffusive_flux1, antidiffusive_flux2 = cache.ContainerAntidiffusiveFlux2D
+  @unpack inverse_weights = dg.basis
+  @unpack positCorrFactor = indicator_IDP
+
+  @unpack var_bounds = indicator_IDP.cache.ContainerShockCapturingIndicator
+
+  var_min = var_bounds[index]
+
+  @threaded for element in eachelement(dg, semi.cache)
+    inverse_jacobian = cache.elements.inverse_jacobian[element]
+    for j in eachnode(dg), i in eachnode(dg)
+      var = variable(get_node_vars(u, equations, dg, i, j, element), equations)
+      if var < 0.0
+        println("Error: safe $variable is not safe. element=$element, node: $i $j, value=$var")
+      end
+
+      # Compute bound
+      var_min[i, j, element] = positCorrFactor * var
+
+      # Real one-sided Zalesak-type limiter
+      # * Zalesak (1979). "Fully multidimensional flux-corrected transport algorithms for fluids"
+      # * Kuzmin et al. (2010). "Failsafe flux limiting and constrained data projections for equations of gas dynamics"
+      # Note: The Zalesak limiter has to be computed, even if the state is valid, because the correction is
+      #       for each interface, not each node
+      Qm = min(0.0, (var_min[i, j, element] - var) / dt)
```

Comment:
```suggestion
      Qm = min(0, (var_min[i, j, element] - var) / dt)
```
This is more general (and should still be type stable)